plan/expression_rewriter.go

```diff
@@ -245,26 +246,197 @@ func (er *expressionRewriter) handleCompareSubquery(v *ast.CompareSubqueryExpr)
 	return v, true
 }
 
+// handleOtherComparableSubq handles the queries like < any, < max, etc. For example, if the query is t.id < any (select s.id from s),
+// it will be rewrote to t.id < (select max(s.id) from s).
+func (er *expressionRewriter) handleOtherComparableSubq(lexpr, rexpr expression.Expression, np LogicalPlan, useMin bool, cmpFunc string, all bool) {
+	funcName := ast.AggFuncMax
+	if useMin {
+		funcName = ast.AggFuncMin
+	}
+	aggFunc := expression.NewAggFunction(funcName, []expression.Expression{rexpr}, false)
+	agg := &Aggregation{
+		baseLogicalPlan: newBaseLogicalPlan(Agg, er.b.allocator),
+		AggFuncs:        []expression.AggregationFunction{aggFunc},
+	}
+	agg.initIDAndContext(er.b.ctx)
+	agg.self = agg
+	addChild(agg, np)
+	aggCol0 := &expression.Column{
+		ColName:  model.NewCIStr("agg_Col_0"),
+		FromID:   agg.id,
+		Position: 0,
+		RetType:  aggFunc.GetType(),
+	}
+	schema := expression.NewSchema([]*expression.Column{aggCol0})
+	agg.SetSchema(schema)
+	cond, _ := expression.NewFunction(cmpFunc, types.NewFieldType(mysql.TypeTiny), lexpr, aggCol0.Clone())
+	er.buildQuantifierPlan(agg, cond, rexpr, all)
+}
+
+// buildQuantifierPlan adds extra condition for any / all subquery.
+func (er *expressionRewriter) buildQuantifierPlan(agg *Aggregation, cond, rexpr expression.Expression, all bool) {
+	isNullFunc, _ := expression.NewFunction(ast.IsNull, types.NewFieldType(mysql.TypeTiny), rexpr.Clone())
+	sumFunc := expression.NewAggFunction(ast.AggFuncSum, []expression.Expression{isNullFunc}, false)
+	countFuncNull := expression.NewAggFunction(ast.AggFuncCount, []expression.Expression{isNullFunc.Clone()}, false)
+	agg.AggFuncs = append(agg.AggFuncs, sumFunc, countFuncNull)
+	posID := agg.schema.Len()
+	aggColSum := &expression.Column{
+		ColName:  model.NewCIStr("agg_col_sum"),
+		FromID:   agg.id,
+		Position: posID,
+		RetType:  sumFunc.GetType(),
+	}
+	aggColCountNull := &expression.Column{
+		ColName:  model.NewCIStr("agg_col_cnt"),
+		FromID:   agg.id,
+		Position: posID + 1,
+		RetType:  countFuncNull.GetType(),
+	}
+	agg.schema.Append(aggColSum)
+	agg.schema.Append(aggColCountNull)
+	if all {
+		// All of the inner record set should not contain null value. So for t.id < all(select s.id from s), it
+		// should be rewrote to t.id < min(s.id) and if(sum(s.id is null) = 0, true, null).
+		hasNotNull, _ := expression.NewFunction(ast.EQ, types.NewFieldType(mysql.TypeTiny), aggColSum.Clone(), expression.Zero)
+		nullChecker, _ := expression.NewFunction(ast.If, types.NewFieldType(mysql.TypeTiny), hasNotNull, expression.One, expression.Null)
+		cond = expression.ComposeCNFCondition(cond, nullChecker)
+		// If the set is empty, it should always return true.
+		checkEmpty, _ := expression.NewFunction(ast.EQ, types.NewFieldType(mysql.TypeTiny), aggColCountNull.Clone(), &expression.Constant{
+			Value:   types.NewDatum(0),
+			RetType: types.NewFieldType(mysql.TypeShort),
```

Comment:
If the two types both are ok,  it's better to use `expression.Zero` instead of these lines.